java/server/src/org/openqa/selenium/grid/sessionqueue/local/LocalNewSessionQueue.java

```diff
@@ -154,21 +157,39 @@ private void retryRequest(HttpRequest request, RequestId requestId) {
   }
 
   @Override
-  public Optional<HttpRequest> poll() {
+  public Optional<HttpRequest> remove(RequestId id) {
     Lock writeLock = lock.writeLock();
     writeLock.lock();
     try {
-      Optional<HttpRequest> request = Optional.ofNullable(sessionRequests.poll());
-      if (request.isPresent()) {
-        HttpRequest httpRequest = request.get();
-        if (hasRequestTimedOut(httpRequest)) {
-          // TODO Fire NewSessionRejectedEvent with timeout message once request id is passed to poll
-          return Optional.empty();
+      // Peek  the deque and check if the request-id matches. Most cases, it would.
+      // If so poll the deque else iterate over the deque and find a match.
+      Optional<SessionRequest> firstSessionRequest =
+          Optional.ofNullable(sessionRequests.peekFirst());
+
+      Optional<HttpRequest> httpRequest = Optional.empty();
+      if (firstSessionRequest.isPresent()) {
+        if (id.equals(firstSessionRequest.get().getRequestId())) {
+          httpRequest = Optional.ofNullable(sessionRequests.pollFirst().getHttpRequest());
```

Comment:
Why don't we get the `HttpRequest` from `firstSessionRequest` if it is `present` and it equals the `id` parameter?

Reply:
Nevermind, `firstSessionRequest` is obtained through `peekFirst`, which does not remove.